fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneOlapScanPartition.java

```diff
@@ -49,65 +56,145 @@
+            }
+        }
+        NereidsSortedPartitionsCacheManager sortedPartitionsCacheManager = Env.getCurrentEnv()
+                .getSortedPartitionsCacheManager();
+        List<Long> manuallySpecifiedPartitions = scan.getManuallySpecifiedPartitions();
+        Map<Long, PartitionItem> idToPartitions;
+        Optional<SortedPartitionRanges<Long>> sortedPartitionRanges = Optional.empty();
+        if (manuallySpecifiedPartitions.isEmpty()) {
+            Optional<SortedPartitionRanges<?>> sortedPartitionRangesOpt
+                    = sortedPartitionsCacheManager.get(table, scan);
+            if (sortedPartitionRangesOpt.isPresent()) {
+                sortedPartitionRanges = (Optional) sortedPartitionRangesOpt;
             }
+            idToPartitions = partitionInfo.getIdToItem(false);
+        } else {
+            Map<Long, PartitionItem> allPartitions = partitionInfo.getAllPartitions();
+            idToPartitions = allPartitions.keySet().stream()
+                    .filter(manuallySpecifiedPartitions::contains)
+                    .collect(Collectors.toMap(Function.identity(), allPartitions::get));
+        }
+        if (filter != null) {
             List<Long> prunedPartitions = PartitionPruner.prune(
                     partitionSlots, filter.getPredicate(), idToPartitions, ctx.cascadesContext,
                     PartitionTableType.OLAP, sortedPartitionRanges);
-            if (prunedPartitions.isEmpty()) {
-                return new LogicalEmptyRelation(
-                        ConnectContext.get().getStatementContext().getNextRelationId(),
-                        filter.getOutput());
+            return prunedPartitions;
+        } else if (!manuallySpecifiedPartitions.isEmpty()) {
+            return Utils.fastToImmutableList(idToPartitions.keySet());
+        } else {
+            return null;
+        }
+    }
+
+    private List<Long> prunePartitionByTabletIds(LogicalOlapScan scan,
+                                                 OlapTable table,
+                                                 List<Long> prunedPartitionsByFilters) {
+        if (scan.getManuallySpecifiedTabletIds().size() == 0
+                || (prunedPartitionsByFilters != null && prunedPartitionsByFilters.isEmpty())) {
+            // `prunedPartitionsByFilters is not null and is empty` means empty partitions after pruner
+            return prunedPartitionsByFilters;
+        }
+
+        Set<Long> selectedPartitions = new LinkedHashSet<>();
+        if (prunedPartitionsByFilters != null) {
+            selectedPartitions.addAll(prunedPartitionsByFilters);
+        }
+
+        Set<Long> manuallySpecifiedTabletIds = ImmutableSet.copyOf(scan.getManuallySpecifiedTabletIds());
+        List<Long> selectPartitionIds = new ArrayList<>();
+        for (Partition partition : table.getPartitions()) {
+            if (!selectedPartitions.isEmpty() && !selectedPartitions.contains(partition.getId())) {
+                continue;
+            }
```

Comment:
remove local variable selectedPartitions and
```suggestion
            if (prunedPartitionsByFilters != null && !prunedPartitionsByFilters.contains(partition.getId()))) {
                continue;
            }
```

Reply:
The type of `prunedPartitionsByFilters ` is `List`. The function `contains()` may cost too much if the list's length is high.